<a href="https://colab.research.google.com/github/vroomgit/FraudDetectionFNN/blob/main/Fraud_Detection_Feedforward_NN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Using a Feedforward Neural Network to detect fraudulent transactions in a simulated financial payments dataset

Link to dataset: https://www.kaggle.com/datasets/ealaxi/paysim1

## Environment prep and data unzipping (This section needed for Google Colab)

In [8]:
from google.colab import drive
drive.mount("/content/drive")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [9]:
gpu_info = !nvidia-smi
gpu_info = '\n'.join(gpu_info)
if gpu_info.find('failed') >= 0:
  print('Select the Runtime > "Change runtime type" menu to enable a GPU accelerator, ')
  print('and then re-execute this cell.')
else:
  print(gpu_info)

Thu Aug  3 10:47:05 2023       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 525.105.17   Driver Version: 525.105.17   CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla V100-SXM2...  Off  | 00000000:00:04.0 Off |                    0 |
| N/A   38C    P0    40W / 300W |    866MiB / 16384MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-------

## Loading and Preparing Data

In [10]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
import torch
from torch.utils.data import Dataset, DataLoader
import torch.optim as optim
import torch.nn as nn
import torch.nn.functional as F
from torchvision import models
import matplotlib.pyplot as plt
from collections import Counter
import time
from time import sleep
!pip install pycm
import pycm
import prettytable

In [11]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
device

device(type='cuda')

In [12]:
df = pd.read_csv('/content/drive/MyDrive/Projects/Finance Dataset Kaggle/FFDS.csv')
df.head()

step      type    amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
0     1   PAYMENT   9839.64  C1231006815       170136.0       160296.36   
1     1   PAYMENT   1864.28  C1666544295        21249.0        19384.72   
2     1  TRANSFER    181.00  C1305486145          181.0            0.00   
3     1  CASH_OUT    181.00   C840083671          181.0            0.00   
4     1   PAYMENT  11668.14  C2048537720        41554.0        29885.86   

      nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  
0  M1979787155             0.0             0.0        0               0  
1  M2044282225             0.0             0.0        0               0  
2   C553264065             0.0             0.0        1               0  
3    C38997010         21182.0             0.0        1               0  
4  M1230701703             0.0             0.0        0               0

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6362620 entries, 0 to 6362619
Data columns (total 11 columns):
 #   Column          Dtype  
---  ------          -----  
 0   step            int64  
 1   type            object 
 2   amount          float64
 3   nameOrig        object 
 4   oldbalanceOrg   float64
 5   newbalanceOrig  float64
 6   nameDest        object 
 7   oldbalanceDest  float64
 8   newbalanceDest  float64
 9   isFraud         int64  
 10  isFlaggedFraud  int64  
dtypes: float64(5), int64(3), object(3)
memory usage: 534.0+ MB


In [14]:
# Percentage of labels = 1
(len(df)-len(df[df['isFraud']==1]))/len(df)

0.9987091795518198

In [15]:
# dropping column 'isFlaggedFraud'
df.drop(columns=['isFlaggedFraud'],inplace=True)

In [16]:
# Number of nulls in dataset
for col in df.columns:
    print(col, df[col].isnull().sum())

step 0
type 0
amount 0
nameOrig 0
oldbalanceOrg 0
newbalanceOrig 0
nameDest 0
oldbalanceDest 0
newbalanceDest 0
isFraud 0


## In order to be fed into a neural network, the four 'object' type columns will be converted to embeddings. Choosing columns that will be encoded and calculating embedding sizes.



In [17]:
# Changing datatypes of 'object' types to 'category'

df["step"] = df["step"].astype("category")
df["type"] = df["type"].astype("category")
df["nameOrig"] = df["nameOrig"].astype("category")
df["nameDest"] = df["nameDest"].astype("category")

In [18]:
# Defining labels
Y = df['isFraud']
df.drop(columns=['isFraud'],inplace=True)
print(len(df),len(Y))

6362620 6362620


In [19]:
# calculating number of categories in 'category' type columns.
embCol_dict = {}
for n,c in df.items():
  if str(df.dtypes[c.name]) == "category":
    if len(c.cat.categories) > 2:
      embCol_dict[c.name] = len(c.cat.categories)
embCol_dict

{'step': 743, 'type': 5, 'nameOrig': 6353307, 'nameDest': 2722362}

Borrowed from https://jovian.ai/aakanksha-ns/shelter-outcome

Below code calculates the size of embeddings based on number of categories in each column. The output size is min(50,(category size+1)/2). In case of step, nameOrig and nameDest, embeddings help in reducing categories to embeddings of length 50

In [20]:
# computing embedding sizes
embedding_sizes = [(n_categories, min(16, (n_categories+1)//2)) for _,n_categories in embCol_dict.items()]
embedding_sizes

[(743, 16), (5, 3), (6353307, 16), (2722362, 16)]

In [21]:
embCol_list = embCol_dict.keys()
len(embCol_list)

4

Encoding text categories in 'category' columns to integer codes using Sklearn's LabelEncoder

In [22]:
for col in df.columns:
  if (df.dtypes[col]) == "category":
    df[col] = LabelEncoder().fit_transform(df[col])

Portioning 70% data for training and 15% each for validation and testing.

In [23]:
np.random.seed(13)
X_train, X_val_test, y_train, y_val_test = train_test_split(df, Y, test_size=0.30)
X_train.head()

step  type     amount  nameOrig  oldbalanceOrg  newbalanceOrig  \
3992863   297     3   10065.90   2897705            0.0            0.00   
5609048   393     3   35667.84   3662289        69610.0        33942.16   
6043451   489     0  408532.84   5627573        16636.0       425168.84   
1219187   132     1  455613.55   3461200       257470.0            0.00   
5325276   373     0  267214.37   3376548        21180.0       288394.37   

         nameDest  oldbalanceDest  newbalanceDest  
3992863   1427538            0.00            0.00  
5609048   1614022            0.00            0.00  
6043451    534679       809226.78       400693.94  
1219187    423423        94728.63       550342.17  
5325276    401669            0.00            0.00

In [24]:
np.random.seed(13)
X_val, X_test, y_val, y_test = train_test_split(X_val_test, y_val_test, test_size=0.50)

In [25]:
# reset indexes to avoid errors in dataloader class
X_train = X_train.reset_index(drop=True)
y_train = y_train.reset_index(drop=True)

X_val = X_val.reset_index(drop=True)
y_val = y_val.reset_index(drop=True)

X_test = X_test.reset_index(drop=True)
y_test = y_test.reset_index(drop=True)



In [26]:
print(len(X_train),len(y_train),len(X_val),len(y_val),len(X_test),len(y_test))

4453834 4453834 954393 954393 954393 954393


Data after encoding 'category' columns.

In [27]:
df.head()

step  type    amount  nameOrig  oldbalanceOrg  newbalanceOrig  nameDest  \
0     0     3   9839.64    757869       170136.0       160296.36   1662094   
1     0     3   1864.28   2188998        21249.0        19384.72   1733924   
2     0     4    181.00   1002156          181.0            0.00    439685   
3     0     1    181.00   5828262          181.0            0.00    391696   
4     0     3  11668.14   3445981        41554.0        29885.86    828919   

   oldbalanceDest  newbalanceDest  
0             0.0             0.0  
1             0.0             0.0  
2             0.0             0.0  
3         21182.0             0.0  
4             0.0             0.0

Defining the Dataset class.

In [28]:
class FraudData(Dataset):
    def __init__(self, df, y, embCol_list):

        self.X1 = df.loc[:,embCol_list].copy().values.astype(np.int64)
        self.X2 = df.drop(columns = embCol_list).copy().values.astype(np.float32)
        self.y = y

    def __len__(self):
        return len(self.y)

    def __getitem__(self, idx):
        return self.X1[idx], self.X2[idx], self.y[idx]

Creating Dataset objects for Train/Validation/Test data

In [29]:
train_data = FraudData(X_train, y_train, embCol_list)
valid_data = FraudData(X_val, y_val, embCol_list)
test_data = FraudData(X_test, y_test, embCol_list)

In [30]:
print(len(train_data),len(valid_data),len(test_data))

4453834 954393 954393


Defining Hyperparameters and Dataloader objects

In [70]:
# Hyperparameters and creating train and validation loader variables

batch_size = 4096
lr = 0.0001
w_decay = 0.00001



trainloader = DataLoader(train_data, batch_size=batch_size,
                                          shuffle=True, num_workers=0)


validloader = DataLoader(valid_data, batch_size=batch_size,
                                         shuffle=False, num_workers=0)

## Building Feedforward Neural Network.



In [71]:
class FraudNet(nn.Module):
    def __init__(self, embedding_sizes, n_cont):
        super().__init__()
        self.embeddings = nn.ModuleList([nn.Embedding(categories, size) for categories,size in embedding_sizes])
        n_emb = sum(e.embedding_dim for e in self.embeddings)
        self.n_emb, self.n_cont = n_emb, n_cont
        self.lin1 = nn.Linear(self.n_emb + self.n_cont, 256)
        self.lin2 = nn.Linear(256, 512)
        self.lin3 = nn.Linear(512, 2)
        self.bn1 = nn.BatchNorm1d(self.n_cont)
        self.bn2 = nn.BatchNorm1d(256)
        self.bn3 = nn.BatchNorm1d(512)
        self.embedding_drop = nn.Dropout(0.6)
        self.drops = nn.Dropout(0.3)


    def forward(self, x_cat, x_cont):
        x = [e(x_cat[:,i]) for i,e in enumerate(self.embeddings)]
        x = torch.cat(x, 1)
        x = self.embedding_drop(x)
        x2 = self.bn1(x_cont)
        x = torch.cat([x, x2], 1)
        x = F.relu(self.lin1(x))
        x = self.drops(x)
        x = self.bn2(x)
        x = F.relu(self.lin2(x))
        x = self.drops(x)
        x = self.bn3(x)
        x = self.lin3(x)
        return x

n_cont variable contains number of non-encoded columns

In [72]:
n_cont = len(df.columns) - len(embCol_list)

In [73]:
net = FraudNet(embedding_sizes, n_cont).to(device=device)
print(net)

FraudNet(
  (embeddings): ModuleList(
    (0): Embedding(743, 16)
    (1): Embedding(5, 3)
    (2): Embedding(6353307, 16)
    (3): Embedding(2722362, 16)
  )
  (lin1): Linear(in_features=56, out_features=256, bias=True)
  (lin2): Linear(in_features=256, out_features=512, bias=True)
  (lin3): Linear(in_features=512, out_features=2, bias=True)
  (bn1): BatchNorm1d(5, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (bn2): BatchNorm1d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (bn3): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (embedding_drop): Dropout(p=0.6, inplace=False)
  (drops): Dropout(p=0.3, inplace=False)
)


In [74]:
# https://stackoverflow.com/questions/49201236/check-the-total-number-of-parameters-in-a-pytorch-model

from prettytable import PrettyTable
def paramCount(model):
    t = PrettyTable(["Modules", "Parameters"])
    total_parameters = 0
    for name, parameter in model.named_parameters():
        if not parameter.requires_grad:
          continue
        params = parameter.numel()
        t.add_row([name, params])
        total_parameters+=params
    print(t)
    print(f"Total Trainable Params: {total_parameters}")
    return total_parameters

paramCount(net)

+---------------------+------------+
|       Modules       | Parameters |
+---------------------+------------+
| embeddings.0.weight |   11888    |
| embeddings.1.weight |     15     |
| embeddings.2.weight | 101652912  |
| embeddings.3.weight |  43557792  |
|     lin1.weight     |   14336    |
|      lin1.bias      |    256     |
|     lin2.weight     |   131072   |
|      lin2.bias      |    512     |
|     lin3.weight     |    1024    |
|      lin3.bias      |     2      |
|      bn1.weight     |     5      |
|       bn1.bias      |     5      |
|      bn2.weight     |    256     |
|       bn2.bias      |    256     |
|      bn3.weight     |    512     |
|       bn3.bias      |    512     |
+---------------------+------------+
Total Trainable Params: 145371355


145371355

In [75]:
criterion = nn.CrossEntropyLoss()


optimizer = optim.Adam(params = net.parameters(), lr=lr, weight_decay= w_decay)#, nesterov=False, *, maximize=False, foreach=None)


In [5]:
root_path = '/content/drive/My Drive/Projects/Finance Dataset Kaggle/'

## Training and Inference



In [77]:
num_epochs = 24


tlossList = []
vlossList = []

np.random.seed(13)
valAccMax = 0
for epoch in range(0,num_epochs):  # loop over the dataset multiple times
    t=time.time()
    #sleep(0.5)
    #print("Epoch: ",epoch+1)
    trAcc = []
    valAcc = []
    train_batch_acc=0
    valid_batch_acc=0
    batch_loss=0
    valid_loss=0
    for i, data in enumerate(trainloader):
        # get the inputs; data is a list of [inputs, labels]
        x1 = data[0].to(device=device)
        x2 = data[1].to(device=device)
        labels = data[2].type(torch.LongTensor).to(device=device)

        # zero the parameter gradients
        optimizer.zero_grad()

        # forward + backward + optimize
        outputs = net(x1, x2)

        loss = criterion(outputs, labels)
        batch_loss+=loss
        loss.backward()
        optimizer.step()

        # Batch Training Accuracy
        batchPred = torch.argmax(outputs, dim=1)
        batchAcc = np.mean(np.array((batchPred==labels).cpu()))
        trAcc.append(batchAcc)

    train_batch_acc = np.mean(trAcc)
        # Validation Accuracy
        #for id, d in enumerate(testloader):

    for i, (x1, x2, y) in enumerate(validloader):
        #print(x1,x2,y)
        #print(i)
        net.eval()
        inpx1 = x1.to(device=device)
        inpx2 = x2.to(device=device)
        lab = y.type(torch.LongTensor).to(device=device)


        np_lab = 0
        with torch.no_grad():
          out = net(inpx1,inpx2)
          vloss = criterion(out, lab)
          valid_loss+=vloss
          np_lab = np.array(lab.cpu())

        valPred = torch.argmax(out, dim=1)
        vAcc = np.mean(np.array((valPred==lab).cpu()))
        valAcc.append(vAcc)

    valid_batch_acc = np.mean(valAcc)
    if valid_batch_acc > valAccMax and epoch > 5:
      valAccMax = valid_batch_acc
      if valid_batch_acc > 0.9999:
        p = root_path + "modelAutoSave_" + "_" + str(round(valAccMax*100000,0)) + "_" + str(epoch) + "_" + str(batch_size) + "_" + str(lr) + "_" + str(w_decay) + "_" + ".pth"
        torch.save(net.state_dict(), p)


    """
    print("\n")
    for l in [0,1]:
      rel_labels = np.array((lab == l).cpu())
      rel_predictions = np.array((valPred == l).cpu())
      a = np.bitwise_and(rel_predictions,rel_labels)
      try:
        lAcc = len(a[a==True])/np.count_nonzero(np_lab == l)
        print("Accuracy for label",l,":",lAcc)
      except:
        print("Division by zero error for label:",l)
    """
    tlossList.append(batch_loss.item()/len(train_data))
    vlossList.append(valid_loss.item()/len(valid_data))


    # print statistics
    print("Epoch",epoch+1," T Loss:", batch_loss.item()/len(train_data), "V Loss:", valid_loss.item()/len(valid_data),\
          "T Accuracy:", round(train_batch_acc,10), "V Accuracy:", round(valid_batch_acc,10),\
          "Max Val Acc:", round(valAccMax,6), "Time", round((time.time()-t)/60,2), "min")


print('Finished Training')

Epoch 1  T Loss: 5.574123828818924e-05 V Loss: 3.875455228766166e-06 T Accuracy: 0.9095569467 V Accuracy: 0.9987302601 Max Val Acc: 0 Time 1.74 min
Epoch 2  T Loss: 1.793813263855686e-06 V Loss: 1.5570574852484645e-06 T Accuracy: 0.9988744399 V Accuracy: 0.9989462306 Max Val Acc: 0 Time 1.78 min
Epoch 3  T Loss: 1.4416970095389632e-06 V Loss: 1.3368570520547504e-06 T Accuracy: 0.9989943723 V Accuracy: 0.999080821 Max Val Acc: 0 Time 1.71 min
Epoch 4  T Loss: 1.1982161025220639e-06 V Loss: 1.1278520570418761e-06 T Accuracy: 0.9991044437 V Accuracy: 0.9991444645 Max Val Acc: 0 Time 1.79 min
Epoch 5  T Loss: 1.0533126937719449e-06 V Loss: 1.0315231499637257e-06 T Accuracy: 0.9991701911 V Accuracy: 0.999204978 Max Val Acc: 0 Time 1.71 min
Epoch 6  T Loss: 9.53912422684985e-07 V Loss: 9.334258880078519e-07 T Accuracy: 0.9992244945 V Accuracy: 0.9992373214 Max Val Acc: 0 Time 1.69 min
Epoch 7  T Loss: 8.641488715629909e-07 V Loss: 8.632192328477767e-07 T Accuracy: 0.9992790222 V Accuracy: 0.

In [78]:
# Save model
model_scripted = torch.jit.script(net)
model_scripted.save(root_path+'model_scripted_50.pt')

In [ ]:
from google.colab import runtime
runtime.unassign()

In [6]:
model = torch.jit.load(root_path+'model_scripted_50.pt')
model.eval()

RecursiveScriptModule(
  original_name=FraudNet
  (embeddings): RecursiveScriptModule(
    original_name=ModuleList
    (0): RecursiveScriptModule(original_name=Embedding)
    (1): RecursiveScriptModule(original_name=Embedding)
    (2): RecursiveScriptModule(original_name=Embedding)
    (3): RecursiveScriptModule(original_name=Embedding)
  )
  (lin1): RecursiveScriptModule(original_name=Linear)
  (lin2): RecursiveScriptModule(original_name=Linear)
  (lin3): RecursiveScriptModule(original_name=Linear)
  (bn1): RecursiveScriptModule(original_name=BatchNorm1d)
  (bn2): RecursiveScriptModule(original_name=BatchNorm1d)
  (bn3): RecursiveScriptModule(original_name=BatchNorm1d)
  (embedding_drop): RecursiveScriptModule(original_name=Dropout)
  (drops): RecursiveScriptModule(original_name=Dropout)
)

In [31]:
# Loading testloader in variables for use in inference in training loop in next cell
testloader = DataLoader(test_data, batch_size=len(test_data),
                                         shuffle=False, num_workers=0)

for i, (x1, x2, y) in enumerate(testloader):
  inpx1_t = x1.to(device=device)
  inpx2_t = x2.to(device=device)
  lab_t = y.type(torch.CharTensor).to(device=device)

Checking accuracy on Testing Set

In [33]:
# Evaluation for test data

with torch.no_grad():
  out = model(inpx1_t,inpx2_t)

a = torch.argmax(out, dim=1)
b = np.mean(np.array((a==lab_t).cpu()))
print("Test Accuracy: ",b,"Correctly Classified Datapoints: ",len(test_data)*b,"Incorrectly Classified Datapoints: ",len(test_data)-len(test_data)*b )

Test Accuracy:  0.9995903155199168 Correctly Classified Datapoints:  954002.0 Incorrectly Classified Datapoints:  391.0


## Generating Confusion Matrix

In [34]:
test_labels = np.array(lab_t.cpu())
test_predictions = np.array(a.cpu())

In [37]:
import locale
locale.getpreferredencoding = lambda: "UTF-8"

In [38]:
!pip install pycm

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 70.3/70.3 kB 1.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 599.4/599.4 kB 11.0 MB/s eta 0:00:00


In [35]:
# Confusion Matrix
import pycm
pycm.ConfusionMatrix(test_labels, test_predictions).print_matrix()#, cmap='Blues', columns = ['regular','fraud'])


Predict      0            1            
Actual
0            953127       45           

1            346          875          




The above confusion matrix suggests the following:
1. Out of 1221 fraud cases in test set, 875 were correctly classified as Fraud (True Negative) and 342 were classified as regular transactions (False Positives). Thus, the model is 68.96% accurate in finding fraud.
2. The number of False Negatives is low (40) which means only a few regular transactions are falsely flagged as fraud.
3. By tuning the number of false positive and false negative cases, a financial organization can determine how much fraud they will allow in order to not bother customers with regular transactions with False Negative flags. In other words, the increase in fraud detection capabilites should not come at the expense of more false negatives.


#TabTransformer Experiment

In [ ]:
import torch
!pip install tab-transformer-pytorch

from tab_transformer_pytorch import FTTransformer

In [ ]:
model = FTTransformer(
    categories = (10, 5, 6, 5, 8),      # tuple containing the number of unique values within each category
    num_continuous = 10,                # number of continuous values
    dim = 32,                           # dimension, paper set at 32
    dim_out = 1,                        # binary prediction, but could be anything
    depth = 6,                          # depth, paper recommended 6
    heads = 8,                          # heads, paper recommends 8
    attn_dropout = 0.1,                 # post-attention dropout
    ff_dropout = 0.1                    # feed forward dropout
)

x_categ = torch.randint(0, 5, (1, 5))     # category values, from 0 - max number of categories, in the order as passed into the constructor above
x_numer = torch.randn(1, 10)              # numerical value

pred = model(x_categ, x_numer) # (1, 1)